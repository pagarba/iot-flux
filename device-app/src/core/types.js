export default {
  LOGIN_REQUEST: 'LOGIN_REQUEST',
  LOGIN_SUCCESS: 'LOGIN_SUCCESS',
  LOGIN_FAILURE: 'LOGIN_FAILURE',
  LOGOUT_SUCCESS: 'LOGOUT_SUCCESS',
  LOGOUT_FAILURE: 'LOGOUT_FAILURE',
  ACCOUNT_FETCH_SUCCESS: 'ACCOUNT_FETCH_SUCCESS',

  CREATE_USER_REQUEST: 'CREATE_USER_REQUEST',
  CREATE_USER_SUCCESS: 'CREATE_USER_SUCCESS',
  CREATE_USER_FAILURE: 'CREATE_USER_FAILURE',

  CREATE_DEVICE_REQUEST: 'CREATE_DEVICE_REQUEST',
  CREATE_DEVICE_SUCCESS: 'CREATE_DEVICE_SUCCESS',
  CREATE_DEVICE_FAILURE: 'CREATE_DEVICE_FAILURE',

  EDIT_DEVICE_REQUEST: 'EDIT_DEVICE_REQUEST',
  EDIT_DEVICE_SUCCESS: 'EDIT_DEVICE_SUCCESS',
  EDIT_DEVICE_FAILURE: 'EDIT_DEVICE_FAILURE',

  GET_DEVICES_REQUEST: 'GET_DEVICES_REQUEST',
  GET_DEVICES_SUCCESS: 'GET_DEVICES_SUCCESS',
  GET_DEVICES_FAILURE: 'GET_DEVICES_FAILURE',

  GET_DEVICE_REQUEST: 'GET_DEVICE_REQUEST',
  GET_DEVICE_SUCCESS: 'GET_DEVICE_SUCCESS',
  GET_DEVICE_FAILURE: 'GET_DEVICE_FAILURE',

  DELETE_DEVICE_REQUEST: 'DELETE_DEVICE_REQUEST',
  DELETE_DEVICE_SUCCESS: 'DELETE_DEVICE_SUCCESS',
  DELETE_DEVICE_FAILURE: 'DELETE_DEVICE_FAILURE',

  CREATE_CHANNEL_REQUEST: 'CREATE_CHANNEL_REQUEST',
  CREATE_CHANNEL_SUCCESS: 'CREATE_CHANNEL_SUCCESS',
  CREATE_CHANNEL_FAILURE: 'CREATE_CHANNEL_FAILURE',

  EDIT_CHANNEL_REQUEST: 'EDIT_CHANNEL_REQUEST',
  EDIT_CHANNEL_SUCCESS: 'EDIT_CHANNEL_SUCCESS',
  EDIT_CHANNEL_FAILURE: 'EDIT_CHANNEL_FAILURE',

  GET_CHANNELS_REQUEST: 'GET_CHANNELS_REQUEST',
  GET_CHANNELS_SUCCESS: 'GET_CHANNELS_SUCCESS',
  GET_CHANNELS_FAILURE: 'GET_CHANNELS_FAILURE',

  GET_CHANNEL_REQUEST: 'GET_CHANNEL_REQUEST',
  GET_CHANNEL_SUCCESS: 'GET_CHANNEL_SUCCESS',
  GET_CHANNEL_FAILURE: 'GET_CHANNEL_FAILURE',

  DELETE_CHANNEL_REQUEST: 'DELETE_CHANNEL_REQUEST',
  DELETE_CHANNEL_SUCCESS: 'DELETE_CHANNEL_SUCCESS',
  DELETE_CHANNEL_FAILURE: 'DELETE_CHANNEL_FAILURE',

  ADD_DEVICE_TO_CHANNEL_REQUEST: 'ADD_DEVICE_TO_CHANNEL_REQUEST',
  ADD_DEVICE_TO_CHANNEL_SUCCESS: 'ADD_DEVICE_TO_CHANNEL_SUCCESS',
  ADD_DEVICE_TO_CHANNEL_FAILURE: 'ADD_DEVICE_TO_CHANNEL_FAILURE',

  DELETE_DEVICE_FROM_CHANNEL_REQUEST: 'DELETE_DEVICE_FROM_CHANNEL_REQUEST',
  DELETE_DEVICE_FROM_CHANNEL_SUCCESS: 'DELETE_DEVICE_FROM_CHANNEL_SUCCESS',
  DELETE_DEVICE_FROM_CHANNEL_FAILURE: 'DELETE_DEVICE_FROM_CHANNEL_FAILURE',

  GET_CHANNEL_DEVICES_REQUEST: 'GET_CHANNEL_DEVICES_REQUEST',
  GET_CHANNEL_DEVICES_SUCCESS: 'GET_CHANNEL_DEVICES_SUCCESS',
  GET_CHANNEL_DEVICES_FAILURE: 'GET_CHANNEL_DEVICES_FAILURE',

  CREATE_EVENT_REQUEST: 'CREATE_EVENT_REQUEST',
  CREATE_EVENT_SUCCESS: 'CREATE_EVENT_SUCCESS',
  CREATE_EVENT_FAILURE: 'CREATE_EVENT_FAILURE',

  EDIT_EVENT_REQUEST: 'EDIT_EVENT_REQUEST',
  EDIT_EVENT_SUCCESS: 'EDIT_EVENT_SUCCESS',
  EDIT_EVENT_FAILURE: 'EDIT_EVENT_FAILURE',

  GET_EVENTS_REQUEST: 'GET_EVENTS_REQUEST',
  GET_EVENTS_SUCCESS: 'GET_EVENTS_SUCCESS',
  GET_EVENTS_FAILURE: 'GET_EVENTS_FAILURE',

  GET_EVENT_REQUEST: 'GET_EVENT_REQUEST',
  GET_EVENT_SUCCESS: 'GET_EVENT_SUCCESS',
  GET_EVENT_FAILURE: 'GET_EVENT_FAILURE',

  DELETE_EVENT_REQUEST: 'DELETE_EVENT_REQUEST',
  DELETE_EVENT_SUCCESS: 'DELETE_EVENT_SUCCESS',
  DELETE_EVENT_FAILURE: 'DELETE_EVENT_FAILURE',
}