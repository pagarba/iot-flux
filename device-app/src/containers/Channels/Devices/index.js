import React, { Component } from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';

import { connect } from 'react-redux';
import ExpansionPanel from '@material-ui/core/ExpansionPanel';
import ExpansionPanelDetails from '@material-ui/core/ExpansionPanelDetails';
import ExpansionPanelSummary from '@material-ui/core/ExpansionPanelSummary';
import Typography from '@material-ui/core/Typography';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';

import Button from '@material-ui/core/Button';
import FormControl from '@material-ui/core/FormControl';
import InputLabel from '@material-ui/core/InputLabel';
import MenuItem from '@material-ui/core/MenuItem';
import Select from '@material-ui/core/Select';

import { withStyles } from "@material-ui/core/styles/index";

import ChannelDevicesTable from '../../../components/Tables/ChannelDevicesTable';
import CreateConfirmDialog from '../../../components/Dialogs/CreateConfirmDialog';
import { addDeviceToChannel, deleteDeviceFromChannel, getChannel, getChannels, getChannelDevices } from '../../../core/actions/channel'
import { getDevices } from '../../../core/actions/device'

const styles = theme => ({
  root: {
    display: 'block'
  },
  button: {
    margin: theme.spacing.unit,
    marginTop: '2em',
    color: '#fff',
  },
  channelDevicesListPanel: {
    backgroundColor: '#52cee8',
  },
  heading: {
    fontSize: theme.typography.pxToRem(15),
    flexShrink: 0,
    fontWeight: 'bold',
  },
  formControl: {
    margin: theme.spacing.unit,
    minWidth: '120px'
  },
  secondaryHeading: {
    fontSize: theme.typography.pxToRem(15),
    color: theme.palette.text.secondary,
  },
});

export class Devices extends Component {
  state = {
    expanded: 'panel1',
    deviceId: '',
    isCreateConfirmDialogOpen: false,
  };

  componentDidMount() {
    this.props.getDevices();
  }

  handleChange = panel => (event, expanded) => {
    this.setState({
      expanded: expanded ? panel : false,
    });
  };

  handleBlur = event => {
    if (!this.state[event.target.name])
      this.setState({ [event.target.name]: null });
  };

  handleInputChange = event => {
    this.setState({ [event.target.name]: event.target.value });
  };

  handleHideCreateConfirmDialog = () => {
    this.setState({
      isCreateConfirmDialogOpen: false,
    })
  }

  navigateToView = () => {
    this.setState({
      isCreateConfirmDialogOpen: false,
      expanded: 'panel1'
    });
  }

  navigateToCreate = () => {
    this.setState({
      isCreateConfirmDialogOpen: false,
      expanded: 'panel2'
    });
  }

  addDeviceToChannel = () => {
    const { deviceId } = this.state;
    const { id: channelId } = this.props.channel;

    if (deviceId) {
      this.props.addDeviceToChannel(
        channelId,
        deviceId
      ).then(() => {
        this.props.getChannel(channelId);
        this.props.getChannelDevices(channelId);
        this.setState({
          deviceId: '',
          isCreateConfirmDialogOpen: true,
        })
      });
    }
  };

  deleteDeviceFromChannel = (deviceId) => {
    const { id } = this.props.channel;

    this.props.deleteDeviceFromChannel(id, deviceId)
      .then(() => {
        this.props.getChannel(id);
      })
      .catch(() => {
        alert('Error while disconnecting device');
      });
  }

  render() {
    const { classes, history, channelDevices } = this.props;
    const { expanded } = this.state;
    const connectedDevices = (channelDevices || {}).things || [];

    let devices = this.props.devices ? this.props.devices.things.filter(dev1 => 0 === connectedDevices.filter(dev2 => dev1.id === dev2.id).length) : [];

    return (
      <div className={classes.root}>
        <ExpansionPanel
          className={classes.channelDevicesListPanel}
          expanded={expanded === 'panel1'}
          onChange={this.handleChange('panel1')}
        >
          <ExpansionPanelSummary expandIcon={<ExpandMoreIcon color="primary"/>}>
            <Typography color="primary" className={classes.heading}>CONNECTED DEVICES</Typography>
          </ExpansionPanelSummary>
          <ExpansionPanelDetails>
            <ChannelDevicesTable
              data={connectedDevices}
              history={history}
              onDeleteDeviceFromChannel={this.deleteDeviceFromChannel}
            />
          </ExpansionPanelDetails>
        </ExpansionPanel>
        <ExpansionPanel expanded={expanded === 'panel2'} onChange={this.handleChange('panel2')}>
          <ExpansionPanelSummary expandIcon={<ExpandMoreIcon />}>
            <Typography className={classes.heading}>CONNECT A NEW DEVICE</Typography>
          </ExpansionPanelSummary>
          <ExpansionPanelDetails className={classes.root}>
            <div>
              <FormControl className={classes.formControl}>
                <InputLabel htmlFor="deviceId">Device</InputLabel>
                <Select
                  value={this.state.deviceId}
                  onChange={this.handleInputChange}
                  inputProps={{
                    name: 'deviceId',
                    id: 'deviceId',
                  }}
                >
                  {
                    devices.map((device) => (
                      <MenuItem value={device.id}>{device.name}</MenuItem>
                    ))
                  }
                </Select>
              </FormControl>
            </div>
            <Button
              className={classes.button}
              color="secondary"
              onClick={this.addDeviceToChannel}
              variant="contained"
              disabled={this.props.isAddingDeviceToChannel}
            >
              Connect Device
            </Button>
          </ExpansionPanelDetails>
        </ExpansionPanel>
        <CreateConfirmDialog
          open={this.state.isCreateConfirmDialogOpen}
          onClickDismiss={this.handleHideCreateConfirmDialog}
          onClickView={this.navigateToView}
          onClickCreate={this.navigateToCreate}
          contentText="Device was added to channel"
          titleText="Confirm"
          viewButtonText="View Devices"
          createButtonText="Connect Device"
        />
      </div>
    )
  }
}

Devices.propTypes = {
  classes: PropTypes.object.isRequired,
  theme: PropTypes.object.isRequired,
};

function mapStateToProps(state) {
  return {
    devices: state.rootReducer.device.devices,
    channelDevices: state.rootReducer.channel.channelDevices,
    isAddingDeviceToChannel: state.rootReducer.channel.isAddingDeviceToChannel,
  }
}

function mapDispatchToProps(dispatch) {
  return {
    addDeviceToChannel: (channelId, deviceId) => dispatch(addDeviceToChannel(channelId, deviceId)),
    deleteDeviceFromChannel: (channelId, deviceId) => dispatch(deleteDeviceFromChannel(channelId, deviceId)),
    getChannels: () => dispatch(getChannels()),
    getChannelDevices: (deviceId) => dispatch(getChannelDevices(deviceId)),
    getDevices: () => dispatch(getDevices()),
    getChannel: (id) => dispatch(getChannel(id)),
  }
}

const WithStyles =  withStyles(styles, { withTheme: true })(Devices);
export default connect(mapStateToProps, mapDispatchToProps)(WithStyles);