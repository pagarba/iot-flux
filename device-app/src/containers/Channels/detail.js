import React, { Component } from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';

import { connect } from 'react-redux';

import { withStyles } from "@material-ui/core/styles/index";
import {getChannel, editChannel, getChannelDevices} from '../../core/actions/channel'
import ChannelDevices from './Devices';

const styles = theme => ({
  root: {
    display: 'block',
  },
  button: {
    margin: theme.spacing.unit,
    marginTop: '2em',
    color: '#fff',
  },
  heading: {
    fontSize: theme.typography.pxToRem(15),
    flexBasis: '33.33%',
    flexShrink: 0,
    fontWeight: 'bold',
  },
  formControl: {
    margin: theme.spacing.unit,
  },
  secondaryHeading: {
    fontSize: theme.typography.pxToRem(15),
    color: theme.palette.text.secondary,
  },
});

export class ChannelDetail extends Component {
  constructor () {
    super();

    this.state = {
      isEditMode: false,
      name: '',
    };
  }

  componentWillMount() {
    const channelId = this.props.match.params.channelId;

    this.props.getChannel(channelId);
    this.props.getChannelDevices(channelId);
  }

  componentWillReceiveProps(props) {
    this.setState({
      name: props.channel.name,
    })
  }

  editChannel = () => {
    const { name } = this.state;
    const channelId = this.props.match.params.channelId;

    this.props.editChannel(channelId, {
      name
    }).then(() => {
      this.props.getChannel(channelId).then(() => {
        this.setState({
          isEditMode: false,
        })
      });
    });
  };

  handleInputChange = event => {
    this.setState({ [event.target.name]: event.target.value });
  };

  render() {
    const { classes } = this.props;

    const channel = this.props.channel || {};

    return (
      <div className={classes.root}>
        <ChannelDevices
          channel={channel}
        />
      </div>
    )
  }
}

ChannelDetail.propTypes = {
  classes: PropTypes.object.isRequired,
  theme: PropTypes.object.isRequired,
};

function mapStateToProps(state) {
  return {
    channel: state.rootReducer.channel.channel,
    isEditingChannel: state.rootReducer.channel.isEditingChannel,
  }
}

function mapDispatchToProps(dispatch) {
  return {
    getChannel: (id) => dispatch(getChannel(id)),
    editChannel: (id, data) => dispatch(editChannel(id, data)),
    getChannelDevices: (deviceId) => dispatch(getChannelDevices(deviceId)),
  }
}

const WithStyles =  withStyles(styles, { withTheme: true })(ChannelDetail);
export default connect(mapStateToProps, mapDispatchToProps)(WithStyles);
