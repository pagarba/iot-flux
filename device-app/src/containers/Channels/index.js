import React, { Component } from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';

import { connect } from 'react-redux';
import qs from 'stringquery';
import ExpansionPanel from '@material-ui/core/ExpansionPanel';
import ExpansionPanelDetails from '@material-ui/core/ExpansionPanelDetails';
import ExpansionPanelSummary from '@material-ui/core/ExpansionPanelSummary';
import Typography from '@material-ui/core/Typography';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';

import Button from '@material-ui/core/Button';
import FormControl from '@material-ui/core/FormControl';
import FormHelperText from '@material-ui/core/FormHelperText';
import Input from '@material-ui/core/Input';
import InputLabel from '@material-ui/core/InputLabel';

import { withStyles } from "@material-ui/core/styles/index";

import ChannelsTable from '../../components/Tables/ChannelsTable';
import CreateConfirmDialog from '../../components/Dialogs/CreateConfirmDialog';
import { createChannel, getChannels, deleteChannel, editChannel } from '../../core/actions/channel'

const styles = theme => ({
  root: {
    display: 'block'
  },
  button: {
    margin: theme.spacing.unit,
    marginTop: '2em',
    color: '#fff',
  },
  channelsListPanel: {
    backgroundColor: '#52cee8',
  },
  heading: {
    fontSize: theme.typography.pxToRem(15),
    flexShrink: 0,
    fontWeight: 'bold',
  },
  formControl: {
    margin: theme.spacing.unit,
  },
  secondaryHeading: {
    fontSize: theme.typography.pxToRem(15),
    color: theme.palette.text.secondary,
  },
});

export class Channels extends Component {
  state = {
    expanded: null,
    channelName: '',
    isCreateConfirmDialogOpen: false,
  };

  componentWillMount() {
    this.props.getChannels();
  }

  componentDidMount() {
    const obj = qs(this.props.location.search);
    this.setState({
      expanded: obj.mode === 'create' ? 'panel2' : 'panel1'
    })
  }

  handleChange = panel => (event, expanded) => {
    this.setState({
      expanded: expanded ? panel : false,
    });
  };

  handleBlur = event => {
    if (!this.state[event.target.name])
      this.setState({ [event.target.name]: null });
  };

  handleHideCreateConfirmDialog = () => {
    this.setState({
      isCreateConfirmDialogOpen: false,
    })
  }

  handleInputChange = event => {
    this.setState({ [event.target.name]: event.target.value });
  };

  navigateToView = () => {
    this.setState({
      isCreateConfirmDialogOpen: false,
      expanded: 'panel1'
    });
  }

  navigateToCreate = () => {
    this.setState({
      isCreateConfirmDialogOpen: false,
      expanded: 'panel2'
    });
  }

  addChannel = () => {
    const { channelName } = this.state;

    if (channelName) {
      this.props.createChannel({
        name: channelName,
      }).then(() => {
        this.props.getChannels();
        this.setState({
          channelName: '',
          isCreateConfirmDialogOpen: true,
        })
      });
    }
  };

  deleteChannel = (id) => {
    this.props.deleteChannel(id)
      .then(() => {
        this.props.getChannels();
      })
      .catch(() => {
        alert('Error while deleting channel');
      });
  }

  render() {
    const { classes, history } = this.props;
    const { expanded } = this.state;

    return (
      <div className={classes.root}>
        <ExpansionPanel
          className={classes.channelsListPanel}
          expanded={expanded === 'panel1'}
          onChange={this.handleChange('panel1')}
        >
          <ExpansionPanelSummary expandIcon={<ExpandMoreIcon color="primary"/>}>
            <Typography color="primary" className={classes.heading}>MY CHANNELS</Typography>
          </ExpansionPanelSummary>
          <ExpansionPanelDetails>
            <ChannelsTable
              history={history}
              data={this.props.channels ? this.props.channels.channels : []}
              onDeleteChannel={this.deleteChannel}
              editChannel={this.props.editChannel}
              getChannels={this.props.getChannels}
            />
          </ExpansionPanelDetails>
        </ExpansionPanel>
        <ExpansionPanel expanded={expanded === 'panel2'} onChange={this.handleChange('panel2')}>
          <ExpansionPanelSummary expandIcon={<ExpandMoreIcon />}>
            <Typography className={classes.heading}>ADD A NEW CHANNEL</Typography>
          </ExpansionPanelSummary>
          <ExpansionPanelDetails className={classes.root}>
            <FormControl className={classes.formControl} error={this.state.channelName === null} fullWidth aria-describedby="component-error-text">
              <InputLabel htmlFor="component-error">Channel Name*</InputLabel>
              <Input name="channelName" value={this.state.channelName} onChange={this.handleInputChange} onBlur={this.handleBlur} />
              <FormHelperText id="component-error-text">This field is required</FormHelperText>
            </FormControl>

            <Button
              className={classes.button}
              color="secondary"
              onClick={this.addChannel}
              variant="contained"
              disabled={this.props.isCreatingChannel}
            >
              Add Channel
            </Button>
          </ExpansionPanelDetails>
        </ExpansionPanel>
        <CreateConfirmDialog
          open={this.state.isCreateConfirmDialogOpen}
          onClickDismiss={this.handleHideCreateConfirmDialog}
          onClickView={this.navigateToView}
          onClickCreate={this.navigateToCreate}
          contentText="A new channel has been created"
          titleText="Confirm"
          viewButtonText="View Channels"
          createButtonText="Create Channel"
        />
      </div>
    )
  }
}

Channels.propTypes = {
  classes: PropTypes.object.isRequired,
  theme: PropTypes.object.isRequired,
};

function mapStateToProps(state) {
  return {
    channels: state.rootReducer.channel.channels,
    isCreatingChannel: state.rootReducer.channel.isCreatingChannel,
  }
}

function mapDispatchToProps(dispatch) {
  return {
    getChannels: () => dispatch(getChannels()),
    createChannel: data => dispatch(createChannel(data)),
    deleteChannel: id => dispatch(deleteChannel(id)),
    editChannel: (id, data) => dispatch(editChannel(id, data)),
  }
}

const WithStyles =  withStyles(styles, { withTheme: true })(Channels);
export default connect(mapStateToProps, mapDispatchToProps)(WithStyles);