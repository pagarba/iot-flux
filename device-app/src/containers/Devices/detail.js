import React, { Component } from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';

import { connect } from 'react-redux';

import FormControl from '@material-ui/core/FormControl';
import Input from '@material-ui/core/Input';

import { withStyles } from "@material-ui/core/styles/index";
import { getDevice, editDevice } from '../../core/actions/device';

const styles = theme => ({
  root: {
    display: 'block'
  },
  button: {
    margin: theme.spacing.unit,
    marginTop: '2em',
    color: '#fff',
  },
  heading: {
    fontSize: theme.typography.pxToRem(15),
    flexBasis: '33.33%',
    flexShrink: 0,
    fontWeight: 'bold',
  },
  formControl: {
    margin: theme.spacing.unit,
  },
  secondaryHeading: {
    fontSize: theme.typography.pxToRem(15),
    color: theme.palette.text.secondary,
  },
});

export class DeviceDetail extends Component {
  constructor () {
    super();

    this.state = {
      isEditMode: false,
      key: '',
      name: '',
      type: '',
    };
  }

  componentWillMount() {
    const deviceId = this.props.match.params.deviceId;
    this.props.getDevice(deviceId);
  }

  componentWillReceiveProps(props) {
    this.setState({
      key: props.device.key,
      name: props.device.name,
      type: props.device.type,
    })
  }

  editDevice = () => {
    const { name, type } = this.state;
    const deviceId = this.props.match.params.deviceId;

    this.props.editDevice(deviceId, {
      name,
      type
    }).then(() => {
      this.props.getDevice(deviceId).then(() => {
        this.setState({
          isEditMode: false,
        })
      });
    });
  }

  handleInputChange = event => {
    this.setState({ [event.target.name]: event.target.value });
  };

  render() {
    const { classes } = this.props;
    const { isEditMode } = this.state;

    const device = this.props.device || {};

    return (
      <div className={classes.root}>
        <table>
          <tbody>
            <tr>
              <td><b>Device ID: </b></td>
              <td>
                {device.id}
              </td>
            </tr>
            <tr>
              <td><b>Device Key: </b></td>
              <td>{ device.key }</td>
            </tr>
            <tr>
              <td><b>Device Type: </b></td>
              <td>{device.type}</td>
            </tr>
            <tr>
              <td><b>Device Name: </b></td>
              <td>
                {
                  isEditMode ? (
                    <FormControl className={classes.formControl} error={this.state.name === null} fullWidth aria-describedby="component-error-text">
                      <Input name="name" value={this.state.name} onChange={this.handleInputChange} />
                    </FormControl>
                  ) : device.name
                }
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    )
  }
}

DeviceDetail.propTypes = {
  classes: PropTypes.object.isRequired,
  theme: PropTypes.object.isRequired,
};

function mapStateToProps(state) {
  return {
    device: state.rootReducer.device.device,
    isEditingDevice: state.rootReducer.device.isEditingDevice,
  }
}

function mapDispatchToProps(dispatch) {
  return {
    getDevice: (id) => dispatch(getDevice(id)),
    editDevice: (id, data) => dispatch(editDevice(id, data)),
  }
}

const WithStyles =  withStyles(styles, { withTheme: true })(DeviceDetail);
export default connect(mapStateToProps, mapDispatchToProps)(WithStyles);
