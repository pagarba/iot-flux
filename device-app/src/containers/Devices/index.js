import React, { Component } from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';

import { connect } from 'react-redux';
import ExpansionPanel from '@material-ui/core/ExpansionPanel';
import ExpansionPanelDetails from '@material-ui/core/ExpansionPanelDetails';
import ExpansionPanelSummary from '@material-ui/core/ExpansionPanelSummary';
import Typography from '@material-ui/core/Typography';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';

import Button from '@material-ui/core/Button';
import FormControl from '@material-ui/core/FormControl';
import FormHelperText from '@material-ui/core/FormHelperText';
import Input from '@material-ui/core/Input';
import InputLabel from '@material-ui/core/InputLabel';
import MenuItem from '@material-ui/core/MenuItem';
import Select from '@material-ui/core/Select';

import { withStyles } from "@material-ui/core/styles/index";

import DevicesTable from '../../components/Tables/DevicesTable';
import CreateConfirmDialog from '../../components/Dialogs/CreateConfirmDialog';
import { createDevice, getDevices, deleteDevice, editDevice } from '../../core/actions/device'
import qs from 'stringquery'

const styles = theme => ({
  root: {
    display: 'block'
  },
  button: {
    margin: theme.spacing.unit,
    marginTop: '2em',
    color: '#fff',
  },
  devicesListPanel: {
    backgroundColor: '#52cee8',
  },
  heading: {
    fontSize: theme.typography.pxToRem(15),
    flexShrink: 0,
    fontWeight: 'bold',
  },
  formControl: {
    margin: theme.spacing.unit,
  },
  secondaryHeading: {
    fontSize: theme.typography.pxToRem(15),
    color: theme.palette.text.secondary,
  },
});

export class Devices extends Component {
  state = {
    expanded: null,
    deviceType: '',
    deviceName: '',
    deviceKey: '',
    isCreateConfirmDialogOpen: false,
  };

  componentWillMount() {
    this.props.getDevices();
  }

  componentDidMount() {
    const obj = qs(this.props.location.search);
    this.setState({
      expanded: obj.mode === 'create' ? 'panel2' : 'panel1'
    })
  }

  handleChange = panel => (event, expanded) => {
    this.setState({
      expanded: expanded ? panel : false,
    });
  };

  handleBlur = event => {
    if (!this.state[event.target.name])
      this.setState({ [event.target.name]: null });
  };

  handleHideCreateConfirmDialog = () => {
    this.setState({
      isCreateConfirmDialogOpen: false,
    })
  }

  handleInputChange = event => {
    this.setState({ [event.target.name]: event.target.value });
  };

  navigateToView = () => {
    this.setState({
      isCreateConfirmDialogOpen: false,
      expanded: 'panel1'
    });
  }

  navigateToCreate = () => {
    this.setState({
      isCreateConfirmDialogOpen: false,
      expanded: 'panel2'
    });
  }

  addDevice = () => {
    const { deviceName, deviceType } = this.state;

    if (deviceName && deviceType) {
      this.props.createDevice({
        type: deviceType,
        name: deviceName,
      }).then(() => {
        this.props.getDevices();
        this.setState({
          deviceName: '',
          deviceType: '',
          isCreateConfirmDialogOpen: true,
        })
      });
    }
  };

  deleteDevice = (id) => {
    this.props.deleteDevice(id)
      .then(() => {
        this.props.getDevices();
      })
      .catch(() => {
        alert('Error while deleting device');
      });
  }

  render() {
    const { classes, history } = this.props;
    const { expanded } = this.state;

    return (
      <div className={classes.root}>
        <ExpansionPanel
          className={classes.devicesListPanel}
          expanded={expanded === 'panel1'}
          onChange={this.handleChange('panel1')}
        >
          <ExpansionPanelSummary expandIcon={<ExpandMoreIcon color="primary"/>}>
            <Typography color="primary" className={classes.heading}>MY DEVICES</Typography>
          </ExpansionPanelSummary>
          <ExpansionPanelDetails>
            <DevicesTable
              history={history}
              data={this.props.devices ? this.props.devices.things : []}
              editDevice={this.props.editDevice}
              getDevices={this.props.getDevices}
              onDeleteDevice={this.deleteDevice}
            />
          </ExpansionPanelDetails>
        </ExpansionPanel>
        <ExpansionPanel expanded={expanded === 'panel2'} onChange={this.handleChange('panel2')}>
          <ExpansionPanelSummary expandIcon={<ExpandMoreIcon />}>
            <Typography className={classes.heading}>ADD A NEW DEVICE</Typography>
          </ExpansionPanelSummary>
          <ExpansionPanelDetails className={classes.root}>
            <FormControl className={classes.formControl} error={this.state.deviceName === null} fullWidth aria-describedby="component-error-text">
              <InputLabel htmlFor="component-error">Device Name*</InputLabel>
              <Input name="deviceName" value={this.state.deviceName} onChange={this.handleInputChange} onBlur={this.handleBlur} />
              <FormHelperText id="component-error-text">This field is required</FormHelperText>
            </FormControl>

            <FormControl className={classes.formControl} fullWidth>
              <InputLabel htmlFor="deviceType">Device Type</InputLabel>
              <Select
                value={this.state.deviceType}
                onChange={this.handleInputChange}
                inputProps={{
                  name: 'deviceType',
                  id: 'deviceType',
                }}
              >
                <MenuItem value="device">Device</MenuItem>
                <MenuItem value="app">App</MenuItem>
              </Select>
            </FormControl>

            <Button
              className={classes.button}
              color="secondary"
              onClick={this.addDevice}
              variant="contained"
              disabled={this.props.isCreatingDevice}
            >
              Add Device
            </Button>
          </ExpansionPanelDetails>
        </ExpansionPanel>
        <CreateConfirmDialog
          open={this.state.isCreateConfirmDialogOpen}
          onClickDismiss={this.handleHideCreateConfirmDialog}
          onClickView={this.navigateToView}
          onClickCreate={this.navigateToCreate}
          contentText="A new device has been created"
          titleText="Confirm"
          viewButtonText="View Devices"
          createButtonText="Create Device"
        />
      </div>
    )
  }
}

Devices.propTypes = {
  classes: PropTypes.object.isRequired,
  theme: PropTypes.object.isRequired,
};

function mapStateToProps(state) {
  return {
    devices: state.rootReducer.device.devices,
    isCreatingDevice: state.rootReducer.device.isCreatingDevice,
  }
}

function mapDispatchToProps(dispatch) {
  return {
    getDevices: () => dispatch(getDevices()),
    createDevice: data => dispatch(createDevice(data)),
    deleteDevice: id => dispatch(deleteDevice(id)),
    editDevice: (id, data) => dispatch(editDevice(id, data)),
  }
}

const WithStyles =  withStyles(styles, { withTheme: true })(Devices);
export default connect(mapStateToProps, mapDispatchToProps)(WithStyles);