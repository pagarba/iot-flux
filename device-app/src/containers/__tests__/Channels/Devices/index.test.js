import React from 'react';
import { createShallow } from '@material-ui/core/test-utils';

import { Devices } from '../../../Channels/Devices';

describe('<ChannelDevices />', () => {
  let shallow;

  beforeAll(() => {
    shallow = createShallow();
  });

  it('should render components', () => {
    const wrapper = shallow(
      <Devices
        classes={{}}
        getDevices={() => {}}
        channel={{}}
        theme={{}}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});