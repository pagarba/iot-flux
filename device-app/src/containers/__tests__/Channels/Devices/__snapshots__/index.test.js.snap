// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChannelDevices /> should render components 1`] = `
<div>
  <WithStyles(ExpansionPanel)
    expanded={true}
    onChange={[Function]}
  >
    <WithStyles(ExpansionPanelSummary)
      expandIcon={
        <pure(ExpandMoreIcon)
          color="primary"
        />
      }
    >
      <WithStyles(Typography)
        color="primary"
      >
        CONNECTED DEVICES
      </WithStyles(Typography)>
    </WithStyles(ExpansionPanelSummary)>
    <WithStyles(ExpansionPanelDetails)>
      <WithStyles(ChannelDevicesTable)
        data={Array []}
        onDeleteDeviceFromChannel={[Function]}
      />
    </WithStyles(ExpansionPanelDetails)>
  </WithStyles(ExpansionPanel)>
  <WithStyles(ExpansionPanel)
    expanded={false}
    onChange={[Function]}
  >
    <WithStyles(ExpansionPanelSummary)
      expandIcon={<pure(ExpandMoreIcon) />}
    >
      <WithStyles(Typography)>
        CONNECT A NEW DEVICE
      </WithStyles(Typography)>
    </WithStyles(ExpansionPanelSummary)>
    <WithStyles(ExpansionPanelDetails)>
      <div>
        <WithStyles(FormControl)>
          <WithStyles(WithFormControlContext(InputLabel))
            htmlFor="deviceId"
          >
            Device
          </WithStyles(WithFormControlContext(InputLabel))>
          <WithStyles(WithFormControlContext(Select))
            inputProps={
              Object {
                "id": "deviceId",
                "name": "deviceId",
              }
            }
            onChange={[Function]}
            value=""
          />
        </WithStyles(FormControl)>
      </div>
      <WithStyles(Button)
        color="secondary"
        onClick={[Function]}
        variant="contained"
      >
        Connect Device
      </WithStyles(Button)>
    </WithStyles(ExpansionPanelDetails)>
  </WithStyles(ExpansionPanel)>
  <WithStyles(ConfirmDialog)
    contentText="Device was added to channel"
    createButtonText="Connect Device"
    onClickCreate={[Function]}
    onClickDismiss={[Function]}
    onClickView={[Function]}
    open={false}
    titleText="Confirm"
    viewButtonText="View Devices"
  />
</div>
`;
